research/conditional/utils/conditional_trainer.py

```diff
@@ -229,8 +255,58 @@ def calculate_loss_and_gradient(self, processed_batch: LLMBatch):
             "losses": losses,
         }
 
-    def _apply_gradient(self):
+    def maybe_report_gradient_norm(self, step: int):
+        if self.logging_interval_heavy > 0 and step % self.logging_interval_heavy == 0:
+            with FSDP.summon_full_params(self.model, with_grads=True):
+                for name, value in self.model.named_parameters():
+                    if value.grad is not None:
+                        eps = 1e-5
+                        grad_norm = torch.linalg.norm(value.grad)
+                        param_norm = torch.linalg.norm(self.model_checkpoint[name])
+                        if self.is_logging_process:
+                            self.logger.report_scalar(
+                                title=f"gradient_norm/{name.replace('.', '/')}",
+                                value=grad_norm,
+                                iteration=step,
+                            )
+                            self.logger.report_scalar(
+                                title=f"scaled_gradient_norm/{name.replace('.', '/')}",
+                                value=grad_norm / (param_norm + eps),
+                                iteration=step,
+                            )
+
+    def will_report_update_norm(self, step: int):
+        return (
+            self.should_log_update_norm
+            and (self.logging_interval_heavy > 0)
+            and (step % self.logging_interval_heavy == 0)
+            and isinstance(self.model, FSDP)
+        )
+
+    def maybe_report_update_norm(self, step: int):
+        if self.will_report_update_norm(step):
+            with FSDP.summon_full_params(self.model, with_grads=False):
+                for name, value in self.model.named_parameters():
+                    eps = 1e-5
+                    update_norm = torch.linalg.norm(
+                        value.detach() - self.model_checkpoint[name]
+                    )
+                    param_norm = torch.linalg.norm(self.model_checkpoint[name])
+                    if self.is_logging_process:
+                        self.logger.report_scalar(
```

Comment:
should be part of `will_report_update_norm`

Reply:
TBH not really - we want the `with FSDP(...)` to run in all processes, but the logging to only be done by the logging process.

Reply:
Yup, found that in the docs. 
How about adding "rank0_only=True", and moving if self.is_logging_process: outside `for`